### PR TITLE
added camera shake and stun to OB.

### PR DESCRIPTION
**Pull request by Cthulhu80**
# About the pull request

Added a camera shake and stun feature based on the user's distance to the OB.

# Explain why it's good for the game

OBs should feel like OBs, and adding this feature just helps make OBs feel like an actual large explosion. 

# Changelog

:cl:
add: added camera shake and stun to OB.
/:cl:

### CODE REVIEW

**Review by morrowwolf — COMMENTED** [review]
I'd like to see this on live before merging. Whenever review is sorted please testmerge.

File: code/modules/cm_marines/orbital_cannon.dm [hunk 1]
```diff
@@ -352,8 +352,38 @@ var/list/ob_type_fuel_requirements
...
 	name = "theoretical orbital ammo"
+	var/max_shake_factor // the max strength of the camera shake on impact.
+	var/max_knockdown_time // the max time a mob can be stunned after ob.
 	var/warhead_kind
 
+/obj/structure/ob_ammo/warhead/proc/handle_ob_shake(turf/target)
+	if(!target)
+		return
+
+	for(var/mob/M in range(30, target))
```
[the remark this answers is not in the document]
Comment (Cthulhu80) [reply code/modules/cm_marines/orbital_cannon.dm]: I think I resolved it? idk.

**Comment by Cthulhu80** [comment]
Fixed the requested changes and should be ready for testing @morrowwolf

File: code/modules/cm_marines/orbital_cannon.dm [hunk 2]
```diff
@@ -419,18 +461,23 @@ var/list/ob_type_fuel_requirements
 	var/datum/cause_data/cause_data = create_cause_data(initial(name), source_mob)
 	cell_explosion(target, clear_power, clear_falloff, EXPLOSION_FALLOFF_SHAPE_LINEAR, null, cause_data) //break shit around
 	sleep(clear_delay)
-	//ACTUALLY BLOW SHIT UP
-	if(!target.density)
+
+	// Explosion if turf is not blocked.
+	if(!is_blocked_turf(target))
```
Comment (ondrej008) [inline code/modules/cm_marines/orbital_cannon.dm:466]: is_blocked_turf is NOT the same as turf.density

It checks for dense atoms on the turf, which means that a lot more things than walls would block OBs now.

File: code/modules/cm_marines/orbital_cannon.dm [hunk 3]
```diff
@@ -493,11 +544,13 @@ var/list/ob_type_fuel_requirements
 		for(var/k = 1 to instant_amount)
 			var/turf/U = pick(turf_list)
 			fire_in_a_hole(U)
+
 		sleep(delay_between_clusters)
 
 /obj/structure/ob_ammo/warhead/cluster/proc/fire_in_a_hole(turf/loc)
 	new /obj/effect/overlay/temp/blinking_laser (loc)
 	addtimer(CALLBACK(GLOBAL_PROC, GLOBAL_PROC_REF(cell_explosion), loc, explosion_power, explosion_falloff, EXPLOSION_FALLOFF_SHAPE_LINEAR, null, create_cause_data(initial(name), source_mob)), 1 SECONDS)
+	addtimer(CALLBACK(src, PROC_REF(handle_ob_shake), loc), 1 SECONDS)
```
Comment (morrowwolf) [inline code/modules/cm_marines/orbital_cannon.dm:553]: This is throwing a bunch of runtimes about callback to a qdeleted object. Gonna need a fix before merge.

Comment (Cthulhu80) [reply code/modules/cm_marines/orbital_cannon.dm]: Should be resolved now.

File: code/datums/autocells/explosion.dm [hunk 4]
```diff
@@ -293,6 +293,52 @@ as having entered the turf.
 		create_shrapnel(epicenter, rand(5,9), , ,/datum/ammo/bullet/shrapnel/light/effect/ver1, explosion_cause_data)
 		create_shrapnel(epicenter, rand(5,9), , ,/datum/ammo/bullet/shrapnel/light/effect/ver2, explosion_cause_data)
 
+	// everything below this line is strictly for handling ob camera shakes
+	// it can be modified in the future to handle other explosion types by interacting with power and falloff instead.
+	if(!warhead_kind)
```
Comment (morrowwolf) [inline code/datums/autocells/explosion.dm:298]: I think all of this should really be handled in OB code. If you're specifying "warhead_kind" you're looking at something that likely should be handled via object oriented programming. Sorry to have you refactor it again 😶

**Review by morrowwolf — CHANGES_REQUESTED** [review]

File: code/datums/autocells/explosion.dm (hunk 4 above)
Comment (Cthulhu80) [reply code/datums/autocells/explosion.dm:298]: It was initially like that, but I moved it over with the expectation someone down the line would want to integrate it with this proc. I'll move it back.

**Comment by Cthulhu80** [comment]
Not exactly sure what happened but I messed up this branch creating a new one.

**Pull request closed without merging**